torchreid/apis/training.py

```diff
@@ -0,0 +1,108 @@
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from copy import deepcopy
+
+import torchreid
+from torchreid.engine import build_engine
+from torchreid.optim import LrFinder
+from scripts.default_config import (lr_finder_run_kwargs,
+                                    lr_scheduler_kwargs, model_kwargs,
+                                    optimizer_kwargs, engine_run_kwargs,
+                                    engine_test_kwargs)
+from torchreid.utils import set_random_seed
+from scripts.script_utils import (build_datamanager, build_auxiliary_model,
+                                  put_main_model_on_the_device)
+
+
+def run_lr_finder(cfg, datamanager, model, optimizer, scheduler, classes,
+                  rebuild_model=True, gpu_num=1, split_models=False):
+    if rebuild_model:
+        tmp_model = model
+    else:
+        tmp_model = deepcopy(model)
+
+    engine = build_engine(cfg, datamanager, tmp_model, optimizer, scheduler, initial_lr=cfg.train.lr)
+    lr_finder = LrFinder(engine=engine, **lr_finder_run_kwargs(cfg))
+    aux_lr = lr_finder.process()
+
+    print(f"Estimated learning rate: {aux_lr}")
+    if cfg.lr_finder.stop_after:
+        print("Finding learning rate finished. Terminate the training process")
+        exit()
+
+    # reload all parts of the training
+    # we do not check classification parameters
+    # and do not get num_train_classes the second time
+    # since it's done above and lr finder cannot change parameters of the datasets
+    cfg.train.lr = aux_lr
+    cfg.lr_finder.enable = False
+    set_random_seed(cfg.train.seed, cfg.train.deterministic)
+    datamanager = build_datamanager(cfg, classes)
+    num_train_classes = datamanager.num_train_pids
+    if rebuild_model:
+        model = torchreid.models.build_model(**model_kwargs(cfg, num_train_classes))
+        num_aux_models = len(cfg.mutual_learning.aux_configs)
+        model, _ = put_main_model_on_the_device(model, cfg.use_gpu, gpu_num, num_aux_models, split_models)
+    optimizer = torchreid.optim.build_optimizer(model, **optimizer_kwargs(cfg))
+    scheduler = torchreid.optim.build_lr_scheduler(optimizer=optimizer,
+                                                    num_iter=datamanager.num_iter,
+                                                    **lr_scheduler_kwargs(cfg))
+
+    return cfg.train.lr
+
+
+def run_training(cfg, datamanager, model, optimizer, scheduler, extra_device_ids, init_lr,
+                 test_before_train=True,
```

Comment:
maybe add this to the config file with True as default to avoid hardcoding this parameter. The condition `if test_before_train` doesn't make sense since it is always True.

Reply:
Or remove this condition at all and do the test before train always

Reply:
` test_before_train` is true only in main.py, in ote task I'd like to prevent this behavior.

Reply:
Generally, this API is supposed to be used by any 3rd party code, so don't look at it from the main.py perspective only.